BACKENDS: Fix libretro preset loading

lephilousophe opened the pull request:
This PR is another try to fix https://forums.scummvm.org/viewtopic.php?p=98550#p98550 like #5140 
Instead of trying several paths, the path from where the shader preset is loaded is used as a base path.
This makes sure that all files are coherent.

For this to work, a function has been added in SearchSet to return the Archive a member pertains to.
In addition, a normalization function has been added to Path object: this avoids lossy transcode to String.
The implementation and tests have been copied from normalizePath function: they both behave the same.

There is still a limitation on Windows: the paths saved in configuration (themepath, iconspath, shader path) all contain backslashes.
The Path objects are created with the / delimiter by default this makes all path above considered as an only path component.
This still works, because Windows seems to accept / and \ as delimiters.

antoniou79 commented:
I've only tested with Windows 10 x64 so far. These are tests I did with the other PR (that is superseded by this one).

I will test on Android and Linux next. 

Windows is "special" because of the backslashes used as delimiters which *may* result in our code *sometimes* treating (parts of) the path as a single chunk "folder" even if it is comprised by multiple subfolders delimited by backslashes).

Anyway, here are the findings so far:
WINDOWS 10 x64: 
Cases to test:
- bilinear.glsp (from archive shaders.dat)
		RESULT: WORKS OK
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" in the same folder)
		RESULT: WORKS OK
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" missing) 
		RESULT: WORKS OK (it picks up dependency "stock.glsl" from archive shaders.dat. Fallback works)
		
- anti-aliasing/aa-shader-4.0.glslp from shaders.dat
	dependency: "shaders/aa-shader-4.0.glsl" is on shaders.dat
	dependency: "../sharpen/shaders/adaptive-sharpen.glsl" is on shaders.dat
		RESULT: WORKS OK
- anti-aliasing/aa-shader-4.0.glslp from FS
		RESULT: FAILS WITH:
		WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
		WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-ext.glslp!

- anti-aliasing/aa-shader-4.0-level2.glslp from extended (downloaded) shaders pack "shaders-20230108.dat"
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on shaders-20230108.dat
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on shaders-20230108.dat
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on shaders.dat (!)
		RESULT: FAILS WITH:
		WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
		WARNING: Failed to load shader anti-aliasing/aa-shader-4.0-level2.glslp!

- anti-aliasing/aa-shader-4.0-level2.glslp from FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS 
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is NOT on the FS (it is in the archive shaders.dat)
		RESULT: FAILS WITH: 
		WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
		WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-level2Ext.glslp!
		
- anti-aliasing/aa-shader-4.0-level2.glslp from FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS 
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on the FS (tried both copying "sharpen/shaders/adaptive-sharpen.glsl" to the parent folder of "anti-aliasing/" and within "anti-aliasing/".
		RESULT: FAILS WITH: 
		WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
		WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-level2Ext.glslp!

Things to consider:
- If there is an external (on the FS) adaptive-sharpen.glsl shouldn't that be used before attempting a search to the archive shaders.dat or shaders-20230108.dat?
	Where should such file be placed?
	The dependency path starts with "../" which indicates that the subfolder path leading to this dependency should start from the parent folder of the selected preset	
- If there is no adaptive-sharpen.glsl on the FS, the archive version for adaptive-sharpen.glsl should be used even if it is in a different archive than the selected preset.

lephilousophe commented:
Thanks for the tests.

> Anyway, here are the findings so far: WINDOWS 10 x64: Cases to test:
> 
> 
>     * bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" missing)
>       RESULT: FAILS WITH:
>       WARNING: LibRetroPipeline::loadPasses: Invalid file path 'D:\Games\ScummVMIcons/stock.glsl'!
>       WARNING: Failed to load shader D:\Games\ScummVMIcons\nearest.glslp!
This is expected.
When picking a shader from the file system it should come with everything it needs.
> 
>     * anti-aliasing/aa-shader-4.0.glslp from FS
>       RESULT: FAILS WITH:
>       WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
>       WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-ext.glslp!
I pushed a fix for it but I am not happy with it (a #ifdef based on platform...).
Going outside the base path slipped over my brain when trying to not use platform specific stuff.
> 
>     * anti-aliasing/aa-shader-4.0-level2.glslp from extended (downloaded) shaders pack "shaders-20230108.dat"
>       dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on shaders-20230108.dat
>       dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on shaders-20230108.dat
>       dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on shaders.dat (!)
>       RESULT: FAILS WITH:
>       WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
>       WARNING: Failed to load shader anti-aliasing/aa-shader-4.0-level2.glslp!
This was not something I thought about when doing the PR initialy.
This is now fixed: we look for the file in the matching container first and then fall back on all archives.
> 
>     * anti-aliasing/aa-shader-4.0-level2.glslp from FS
>       dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
>       dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS
>       dependency "../sharpen/shaders/adaptive-sharpen.glsl" is NOT on the FS (it is in the archive shaders.dat)
>       RESULT: FAILS WITH:
>       WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
>       WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-level2Ext.glslp!
Same case as anti-aliasing/aa-shader-4.0.glslp from FS.
> 
>     * anti-aliasing/aa-shader-4.0-level2.glslp from FS
>       dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
>       dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS
>       dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on the FS (tried both copying "sharpen/shaders/adaptive-sharpen.glsl" to the parent folder of "anti-aliasing/" and within "anti-aliasing/".
>       RESULT: FAILS WITH:
>       WARNING: LibRetroPipeline::loadPasses: Invalid file path 'sharpen/shaders/adaptive-sharpen.glsl'!
>       WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-level2Ext.glslp!
Same case as anti-aliasing/aa-shader-4.0.glslp from FS.
> 
> Things to consider:
> 
>     * If there is an external (on the FS) adaptive-sharpen.glsl shouldn't that be used before attempting a search to the archive shaders.dat or shaders-20230108.dat?
Only when using a preset from the FS. This avoids breakage because some file is available on the FS and doesn't match what is expected.
>       Where should such file be placed?
>       The dependency path starts with "../" which indicates that the subfolder path leading to this dependency should start from the parent folder of the selected preset
Yes, it was an overlook on my side... Obviously, on Linux it worked when tested.
> 
>     * If there is no adaptive-sharpen.glsl on the FS, the archive version for adaptive-sharpen.glsl should be used even if it is in a different archive than the selected preset.
I don't agree with that. FS should stay in FS and archives should stay in archives.
As we use archives in an additional way, I added a fallback to search in archives following the current one.

antoniou79 commented:
// LINUX
- bilinear.glsp (from archive shaders.dat)
		RESULT: WORKS OK
		
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" in the same folder)
		RESULT: WORKS OK
		
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" missing)
		RESULT: FAILS WITH:
		WARNING: LibRetroPipeline::loadPasses: Invalid file path '/media/antoniou/BlitzXT/Games/ScummVMIcons/stock.glsl'!
		WARNING: Failed to load shader /media/antoniou/BlitzXT/Games/ScummVMIcons/nearest.glslp!

- anti-aliasing/aa-shader-4.0.glslp from shaders.dat
	dependency: "shaders/aa-shader-4.0.glsl" is on shaders.dat
	dependency: "../sharpen/shaders/adaptive-sharpen.glsl" is on shaders.dat
		RESULT: WORKS OK
		
- anti-aliasing/aa-shader-4.0.glslp from FS
	dependency: "shaders/aa-shader-4.0.glsl" is on FS
	dependency: "../sharpen/shaders/adaptive-sharpen.glsl" missing from FS
		RESULT: FAILS WITH:
	WARNING: LibRetroPipeline::loadPasses: Invalid file path '/media/antoniou/BlitzXT/Games/ScummVMIcons/sharpen/shaders/adaptive-sharpen.glsl'!
	WARNING: Failed to load shader /media/antoniou/BlitzXT/Games/ScummVMIcons/anti-aliasing/aa-shader-4.0-ext.glslp!
	
- anti-aliasing/aa-shader-4.0.glslp from FS
	dependency: "shaders/aa-shader-4.0.glsl" is on FS
	dependency: "../sharpen/shaders/adaptive-sharpen.glsl" is on FS
		RESULT: WORKS OK

- anti-aliasing/aa-shader-4.0-level2.glslp from extended (downloaded) shaders pack "shaders-20230108.dat"
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on shaders-20230108.dat
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on shaders-20230108.dat
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on shaders.dat (!)
		RESULT: WORKS OK!

- anti-aliasing/aa-shader-4.0-level2.glslp from FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is NOT on the FS (it is in the archive shaders.dat)
		RESULT: FAILS WITH:
		WARNING: LibRetroPipeline::loadPasses: Invalid file path '/media/antoniou/BlitzXT/Games/ScummVMIcons/sharpen/shaders/adaptive-sharpen.glsl'!
		WARNING: Failed to load shader /media/antoniou/BlitzXT/Games/ScummVMIcons/anti-aliasing/aa-shader-4.0-level2Ext.glslp!

- anti-aliasing/aa-shader-4.0-level2.glslp from FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on the FS
		RESULT: WORKS OK! (If "sharpen/shaders/adaptive-sharpen.glsl" is copied to the parent folder of "anti-aliasing/" as expected)

antoniou79 commented:
After the latest commit, test results on Windows. All the failures cases are now intended/expected, so this works ok on Windows. Good work!

Also the error messages on the failed cases show the full path for the missing file (which is better).

// WINDOWS: 
- bilinear.glsp (from archive shaders.dat)
		RESULT: WORKS OK
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" in the same folder)
		RESULT: WORKS OK
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" missing)
		RESULT: FAILS (INTENDED/EXPECTED):
		WARNING: LibRetroPipeline::loadPasses: Invalid file path 'D:/Games/ScummVMIcons/stock.glsl'!
		WARNING: Failed to load shader D:\Games\ScummVMIcons\bilinear.glslp!

- anti-aliasing/aa-shader-4.0.glslp from shaders.dat
	dependency: "shaders/aa-shader-4.0.glsl" is on shaders.dat
	dependency: "../sharpen/shaders/adaptive-sharpen.glsl" is on shaders.dat
		RESULT: WORKS OK

- anti-aliasing/aa-shader-4.0.glslp from FS
	dependency: "shaders/aa-shader-4.0.glsl" is on FS
	dependency: "../sharpen/shaders/adaptive-sharpen.glsl" is on FS
		RESULT: WORKS OK

- anti-aliasing/aa-shader-4.0.glslp from FS
	dependency: "shaders/aa-shader-4.0.glsl" is on FS
	dependency: "../sharpen/shaders/adaptive-sharpen.glsl" missing from FS
		RESULT: FAILS (INTENDED/EXPECTED):
		WARNING: LibRetroPipeline::loadPasses: Invalid file path 'D:/Games/ScummVMIcons/sharpen/shaders/adaptive-sharpen.glsl'!
		WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-ext.glslp!

- anti-aliasing/aa-shader-4.0-level2.glslp from extended (downloaded) shaders pack "shaders-20230108.dat"
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on shaders-20230108.dat
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on shaders-20230108.dat
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on shaders.dat (!)
		RESULT: WORKS OK

- anti-aliasing/aa-shader-4.0-level2.glslp from FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" missing from FS (it is in the archive shaders.dat)
		RESULT: FAILS (INTENDED/EXPECTED):
		WARNING: LibRetroPipeline::loadPasses: Invalid file path 'D:/Games/ScummVMIcons/sharpen/shaders/adaptive-sharpen.glsl'!
		WARNING: Failed to load shader D:\Games\ScummVMIcons\anti-aliasing\aa-shader-4.0-level2Ext.glslp!
		
- anti-aliasing/aa-shader-4.0-level2.glslp from FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass1.glsl" is on the FS
	dependency: "shaders/aa-shader-4.0-level2/aa-shader-4.0-level2-pass2.glsl" is on the FS
	dependency "../sharpen/shaders/adaptive-sharpen.glsl" is on the FS ("sharpen/shaders/adaptive-sharpen.glsl" was copied to the parent folder of "anti-aliasing/" and within "anti-aliasing/")
		RESULT: WORKS OK

antoniou79 commented:
Results from Android 12:

// ANDROID
- bilinear.glsp (from archive shaders.dat)
		RESULT: WORKS OK
		
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" in the same folder)
		RESULT: WORKS OK
		
- bilinear.glsp (from FS "Pick file instead..." with dependency "stock.glsl" missing)
		RESULT: FAILS (EXPECTED):
		WARNING: LibRetroPipeline::loadPasses: Invalid file path '/saf/5C5B-B66B%3Agames/ScummvmIcons/stock.glsl'!
		WARNING: Failed to load shader /saf/5C5B-B66B%3Agames/ScummvmIcons/nearest.glslp!

All the anti-aliasing preset test cases fail but for unrelated reason, ie. due to many shader compilation errors, eg. "Could not compile shader sharpen/shaders/adaptive-sharpen.glsl.fragment: ERROR: 0:169: 'texture2D' :   type is for Vulkan api only ")

I will try to find a similar preset for testing that does not have compilation issues on Android

lephilousophe commented:
Thanks, merging now.
